### ISSUE TITLE
Poor performance when using two or more `DirectionalLight`s with different settings

### ISSUE THREAD

**Issue by alice-i-cecile** [bevyengine/bevy#8129]
## Bevy version

0.10

## \[Optional\] Relevant system information

> 2023-03-19T14:12:24.865565Z  INFO bevy_render::renderer: AdapterInfo { name: "NVIDIA GeForce GTX 970", vendor: 4318, device: 5058, device_type: DiscreteGpu, driver: "NVIDIA", driver_info: "525.85.05", backend: Vulkan }
> 2023-03-19T14:12:24.974571Z  INFO bevy_diagnostic::system_information_diagnostics_plugin::internal: SystemInfo { os: "Linux 22.04 Pop!_OS", kernel: "6.2.0-76060200-generic", cpu: "AMD Ryzen 5 5600G with Radeon Graphics", core_count: "6", memory: "15.4 GiB" }

## What you did

1. Spawn a scene with objects and two directional lights with shadows enabled.
2. Either a) set one of the lights to `Visibility::Hidden` or b) set one of the lights to `shadows_enabled: false`.

Note that this doesn't work if you change *both* of the lights, or only have one light.

## What went wrong

FPS is massively degraded.

## Additional information

Encountered in https://github.com/Leafwing-Studios/Emergence/pull/542; feel free to pull and build there if you can't reproduce more simply.

**Comment by alice-i-cecile** [comment]
This smells an awful lot like pessimal branching behavior somehow.

**Comment by superdump** [comment]
Do you mean that with two directional lights where shadow mapping is enabled in both, one is visible and the other not, that performance is worse than with both visible?

Or is your point that having two directional lights where one is visible and has shadow mapping enabled and the other is not visible or does not have shadow mapping enabled should have the same performance as one directional light? (I agree that that should be the case regardless.)

**Comment by alice-i-cecile** [comment]
The former is what I'm observing. Frame times are about 10x longer in that setup.

**Comment by superdump** [comment]
Interesting indeed. That is also bad.

**Comment by alice-i-cecile** [comment]
Yep, lemme try my hand at that. I'll put together a simpler reproducer to reduce the fuss involved I think.

**Comment by alice-i-cecile** [comment]
I've made a simple reproduction based on `many_cubes` without a moving camera.

Couldn't reproduce there :/ I'll go back to my main project and grab those traces.

| Visible | Shadows | Frametime (ms) |
|---------|---------|----------------|
| Neither | Neither | 5.2            |
| One     | Neither | 5.3            |
| Both    | Neither | 5.6            |
| Neither | One     | 5.3            |
| One     | One     | 11             |
| Both    | One     | 98             |
| Neither | Both    | 5.5            |
| One     | Both    | 96             |
| Both    | Both    | 130            |

**Comment by alice-i-cecile** [comment]
Okay, attempting more reproduction: even on my game, I can't reproduce this simply by disabling one light (or its shadows) from the start. It seems to be related to disabling it *later*, after it's been initialized with shadows on.

**Comment by alice-i-cecile** [comment]
Alright, I've reproduced, this time manifesting as a hitch instead (maybe due to RAM stuff on my side??).

![image](https://user-images.githubusercontent.com/3579909/226187883-eb06151d-bff7-41bf-923b-4c52c97c22e6.png)

There are two anomalies: 

1. `prepare_lights` takes an extremely long time.
2. `submit_graph_commands` takes even longer.

**Comment by superdump** [comment]
I tested the commit locally and tried to reproduce but was unable as far as I could tell. Frame rate remained consistent.

I also understood what the code is doing and read through prepare_lights and check_light_mesh_visibility and nothing jumped out at me as being something that would cause this problem.

**Comment by alice-i-cecile** [comment]
I also looked into this more with Rob's help. You could see the processer usage go up in realtime at the transition, but there was no sudden jump in memory usage and there was plenty free.

**Comment by alice-i-cecile** [comment]
VRAM usage on the other hand is very high to start with (thanks huge shadow textures lol), and runs out when toggling.
